Update loader documentation. (#685)
KhronosGroup/Vulkan-Loader: docs/LoaderImplementationInterface.md

<!-- markdownlint-disable MD041 -->
[![Khronos Vulkan][1]][2]

[1]: https://vulkan.lunarg.com/img/Vulkan_100px_Dec16.png "https://www.khronos.org/vulkan/"
[2]: https://www.khronos.org/vulkan/

# Implementation interface to the Vulkan Loader
[![Creative Commons][3]][4]

<!-- Copyright &copy; 2015-2021 LunarG, Inc. -->

[3]: https://i.creativecommons.org/l/by-nd/4.0/88x31.png "Creative Commons License"
[4]: https://creativecommons.org/licenses/by-nd/4.0/

## Table of Contents

* [Overview](#overview)
* [Implementation Discovery](#implementation-discovery)

>   * [Overriding the Default Implementation Usage](#overriding-the-default-implementation-usage)

Reviewer: ```suggestion
  * [Overriding the Default Implementation Discovery](#overriding-the-default-implementation-discovery)
```

For the complete overview of all sections of the loader, please refer to the
[LoaderInterfaceArchitecture.md](LoaderInterfaceArchitecture.md) file.

**NOTE:** While many of the interfaces still use the "icd" sub-string to
identify various behavior associated with implementations, this is purely
historical and should not indicate that the implementing code do so through
the traditional ICD interface.
Granted, the majority of implementations to this date are ICD implementations
targeting specific GPU hardware.

## Implementation Discovery

Vulkan allows multiple implementations each with one or more devices
(represented by a Vulkan `VkPhysicalDevice` object) to be used collectively.
The loader is responsible for discovering available Vulkan implementations on
the system.
Given a list of available implementations, the loader can enumerate all the
physical devices available for an application and return this information to the
application.
The process in which the loader discovers the available implementations on a
system is platform-dependent.
Windows, Linux, Android, and macOS implementation discovery details are listed
below.

> ### Overriding the Default Implementation Usage

Reviewer: ```suggestion
### Overriding the Default Implementation Discovery
```
Usage doesn't really describe what is being overridden here.

```

This is an example which is using the `VK_ICD_FILENAMES` override on macOS to
point to an installation and build of the MoltenVK GitHub repository that
contains the MoltenVK implementation.

See the
[Table of Debug Environment Variables](LoaderInterfaceArchitecture.md#table-of-debug-environment-variables)
in the [LoaderInterfaceArchitecture.md document](LoaderInterfaceArchitecture.md)
for more details

### Implementation Manifest File Usage

As with layers, on Windows, Linux and macOS systems, JSON-formatted manifest
files are used to store implementation information.
In order to find system-installed drivers, the Vulkan loader will read the JSON
files to identify the names and attributes of each driver.
Notice that implementation Manifest files are much simpler than the corresponding
layer Manifest files.

See the
[Current Implementation Manifest File Format](#implementation-manifest-file-format)
section for more details.

### Implementation Discovery on Windows

In order to find available implementations (including installed ICDs), the
loader scans through registry keys specific to Display Adapters and all Software

> Components associated with these adapters
> for the locations of JSON manifest files.

Reviewer: ```suggestion
Components associated with these adapters for the locations of JSON manifest files.
```
(just deleting a stray line-return.

These keys are located in device keys created during driver installation and
contain configuration information for base settings, including OpenGL and

> Direct3D ICD locations.

Reviewer: ```suggestion
Direct3D locations.
```

HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Khronos\Vulkan\Drivers
```

Every implementation in these locations should be given as a DWORD, with value
0, where the name of the value is the full path to a JSON manifest file.
The Vulkan loader will attempt to open each manifest file to obtain the
information about an implementation's shared library (".dll") file.

For example, let us assume the registry contains the following data:

```
[HKEY_LOCAL_MACHINE\SOFTWARE\Khronos\Vulkan\Drivers\]

"C:\vendor a\vk_vendora.json"=dword:00000000
"C:\windows\system32\vendorb_vk.json"=dword:00000001
"C:\windows\system32\vendorc_icd.json"=dword:00000000
```

In this case, the loader will step through each entry, and check the value.
If the value is 0, then the loader will attempt to load the file.
In this case, the loader will open the first and last listings, but not the
middle.
This is because the value of 1 for vendorb_vk.json disables the driver.

The Vulkan loader will open each enabled manifest file found to obtain the name
or pathname of an implementation's shared library (".DLL") file.

Implementations should use the registry locations from the PnP Configuration
Manager wherever practical.

> Typically, this is most important for ICDs, and the location clearly ties the
> ICD to a given device.
> The `SOFTWARE\Khronos\Vulkan\Drivers` location is the older method for locating
> ICDs, but is primary for non-ICD implementations.

Reviewer: Stray references to ICD's
```suggestion
Typically, this is most important for implementations, and the location clearly ties the
implementations to a given device.
The `SOFTWARE\Khronos\Vulkan\Drivers` location is the older method for locating
implementations, but is the primary location for software based implementations.
```

See the
[Implementation Manifest File Format](#implementation-manifest-file-format)
section for more details.

### Implementation Discovery on Linux

On Linux, the Vulkan loader will scan for implementation manifest files using
environment variables or corresponding fallback values if the corresponding
environment variable is not defined:

<table style="width:100%">
  <tr>
    <th>Search Order</th>
    <th>Directory/Environment Variable</th>
    <th>Fallback</th>
    <th>Additional Notes</th>
  </tr>
  <tr>
    <td>1</td>
    <td>$XDG_CONFIG_HOME</td>

>     <td>$HOME/config</td>

Reviewer 2: ```suggestion
    <td>$HOME/.config</td>
```

    <td><b>This path is ignored when running an application with root or

>         super-user access</b>.<br/>

Author: Thanks for the clarification.  We're using secure_getenv based on requests from Chromium, so it sounds like my wording is wrong and should be what you suggest in the first sentence.

>         super-user access</b>.<br/>

Reviewer 2: Am I right in thinking that every time Vulkan-Loader looks at an environment variable, including `$XDG_*`, `$VULKAN_*` and `$HOME`, it's using `secure_getenv()` or equivalent?

If yes, then it might be better to say that *once*, as a general statement about environment variables, rather than putting it in a subset of the places where environment variables are mentioned (and perhaps making people think that other environment variable references are done unconditionally)?

>         super-user access</b>.<br/>

Author: I do have a global mention of it as well as individual locations.  The intent was people tend to focus solely on the sections they're looking at and by mentioning it multiple times, they may see the comment in one of those locations.  If you look in the top-level LoaderInterfaceArchitecture.md file, you'll see that there's now a "Elevated Privileges Caveats" section.  I can try to specifically call out the XDG and other items there if you believe it's useful.

        This is done to ensure that an application with root access does not
        run layers that did not need root access to install.
    </td>
  </tr>
  <tr>
    <td>1</td>
    <td>$XDG_CONFIG_DIRS</td>
    <td>/etc/xdg</td>
    <td></td>
  </tr>
  <tr>
    <td>2</td>
    <td>$SYSCONFDIR</td>
    <td>/etc</td>
    <td>Compile-time option set to possible location of implementations
        installed from non-Linux-distribution-provided packages.
    </td>
  </tr>
  <tr>
    <td>3</td>

>     <td>$EXTRASYSCONFDIR</td>

Reviewer 2: Similarly I'd write this as

```suggestion
    <td>EXTRASYSCONFDIR</td>
```

        installed from non-Linux-distribution-provided packages.
        Typically only set if $SYSCONFDIR is set to something other than /etc
    </td>
  </tr>
  <tr>
    <td>4</td>
    <td>$XDG_DATA_HOME</td>
    <td>$HOME/.local/share</td>
    <td><b>This path is ignored when running an application with root or
        super-user access</b>.<br/>
        This is done to ensure that an application with root access does not
        run layers that did not need root access to install.
    </td>
  </tr>
  <tr>
    <td>5</td>
    <td>$XDG_DATA_DIRS</td>
    <td>/usr/local/share/:/usr/share/</td>
    <td></td>
  </tr>
</table>

The directory lists are concatenated together using the standard platform path
separator (:).
The loader then selects each path, and applies the "/vulkan/icd.d" suffix onto
each and looks in that specific folder for manifest files.

The Vulkan loader will open each manifest file found to obtain the name or
pathname of an implementation's shared library (".dylib") file.

**NOTE** While the order of folders searched for manifest files is well
defined, the order contents are read by the loader in each directory is

> [random due to the behavior of readdir](https://www.ibm.com/support/pages/order-directory-contents-returned-calls-readdir).

Reviewer 2: I wonder whether it would be worthwhile to sort them lexicographically (`strcmp()` order) just to be able to say that there is *some* defined order?

Author: We have had complaints about attempting to do that, unfortunately.  Since that would mean some vendors would always appear above others.  We're working on other methods to provide a more balanced sorting of available devices that are returned from the implementations themselves.

to debug the issue.
This forces every dynamic library's symbols to be fully resolved on load.
If there is a problem with an implementation missing symbols on the current
system, this will expose it and cause the Vulkan loader to fail on loading
the implementation.
It is recommended that `LD_BIND_NOW` along with `VK_LOADER_DEBUG=error,warn`
to expose any issues.

### Using Pre-Production ICDs or Software Implementations

Both software and pre-production ICDs can use an alternative mechanism to
detect their implementations.
Independent Hardware Vendor (IHV) may not want to fully install a pre-production
ICD and so it can't be found in the standard location.
For example, a pre-production ICD may simply be a shared library in the
developer's build tree.
In this case, there should be a way to allow developers to point to such an
ICD without modifying the system-installed ICD(s) on their system.

This need is met with the use of the `VK_ICD_FILENAMES` environment variable,
which will override the mechanism used for finding system-installed
implementations.

In other words, only the implementations listed in `VK_ICD_FILENAMES` will be
used.

See

> [Overriding the Default Implementation Usage](#overriding-the-default-implementation-usage)

Reviewer: ```suggestion
[Overriding the Default Implementation Discovery](#overriding-the-default-implementation-discovery)
```

by simply not exposing the above `vkCreateXXXSurfaceKHR` and
`vkDestroySurfaceKHR` functions.

#### Loader Version 2 Interface Requirements

Version 2 interface has requirements in three areas:
 1. Implementation Vulkan entry-point discovery,
 2. `KHR_surface` related requirements in the WSI extensions,
 3. Vulkan dispatchable object creation requirements.

#### Loader Versions 0 and 1 Interface Requirements

Version 0 and 1 interfaces do not support version negotiation via
`vk_icdNegotiateLoaderICDInterfaceVersion`.
Implementations can distinguish version 0 and version 1 interfaces as follows:
if the loader calls `vk_icdGetInstanceProcAddr` first it supports version 1;
otherwise the loader only supports version 0.

Version 0 interface does not support `vk_icdGetInstanceProcAddr`.
Version 0 interface requirements for obtaining implementation Vulkan
entry-points are as follows:

- The function `vkGetInstanceProcAddr` **must be exported** in the
implementation library and returns valid function pointers for all the Vulkan
API entry points.
- `vkCreateInstance` **must be exported** by the implementation library.
- `vkEnumerateInstanceExtensionProperties` **must be exported** by the

> implementation library.

Author: It wasn't hard.  So I went ahead and did this.
